internal/initializer/crds_migrator.go

```diff
@@ -0,0 +1,112 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package initializer
+
+import (
+	"context"
+
+	extv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+)
+
+// NewCoreCRDsMigrator returns a new *CoreCRDsMigrator.
+func NewCoreCRDsMigrator(crdName, sourceVersion string) *CoreCRDsMigrator {
+	c := &CoreCRDsMigrator{
+		crdName:    crdName,
+		oldVersion: sourceVersion,
+	}
+	return c
+}
+
+// CoreCRDsMigrator makes sure the CRDs are using the latest storage version.
+type CoreCRDsMigrator struct {
+	crdName    string
+	oldVersion string
+}
+
+// Run applies all CRDs in the given directory.
+func (c *CoreCRDsMigrator) Run(ctx context.Context, kube client.Client) error { //nolint:gocyclo // TODO(phisco) refactor
+	var crd extv1.CustomResourceDefinition
+	if err := kube.Get(ctx, client.ObjectKey{Name: c.crdName}, &crd); err != nil {
+		if kerrors.IsNotFound(err) {
+			// nothing to do
+			return nil
+		}
+		return errors.Wrapf(err, "cannot get %s crd", c.crdName)
+	}
+	// no old version in the crd, nothing to do
+	if !sets.NewString(crd.Status.StoredVersions...).Has(c.oldVersion) {
+		return nil
+	}
+	// we need to patch all resources to the new storage version
+	var storageVersion string
+	for _, v := range crd.Spec.Versions {
+		if v.Storage {
+			storageVersion = v.Name
+			break
+		}
+	}
+	var resources = unstructured.UnstructuredList{}
+	resources.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   crd.Spec.Group,
+		Version: storageVersion,
+		Kind:    crd.Spec.Names.ListKind,
+	})
+	var continueToken string
+	for {
+		if err := kube.List(ctx, &resources,
+			client.Limit(500),
+			client.Continue(continueToken),
+		); err != nil {
+			return errors.Wrapf(err, "cannot list %s", resources.GroupVersionKind().String())
+		}
+		for i := range resources.Items {
+			// apply empty patch for storage version upgrade
+			res := resources.Items[i]
+			if err := kube.Patch(ctx, &res, client.RawPatch(types.MergePatchType, []byte(`{}`))); err != nil {
+				return errors.Wrapf(err, "cannot patch %s %q", crd.Spec.Names.Kind, res.GetName())
+			}
+		}
+		continueToken = resources.GetContinue()
+		if continueToken == "" {
+			break
+		}
+	}
+
+	origCrd := crd.DeepCopy()
+	crd.Status.StoredVersions = []string{storageVersion}
+	if err := kube.Status().Patch(ctx, &crd, client.MergeFrom(origCrd)); err != nil {
```

Comment:
@phisco this must be `MergeFromWithOptions(origCrd, client.MergeFromWithOptimisticLock)`. Otherwise, we risk data loss.